packages/rich-text/src/to-tree.js

```diff
@@ -33,27 +10,73 @@ export function toTree( value, multilineTag, settings ) {
 		appendText,
 		onStartIndex,
 		onEndIndex,
-		onEmpty,
 	} = settings;
 	const { formats, text, start, end } = value;
 	const formatsLength = formats.length + 1;
-	const tree = createEmpty( tag );
+	const tree = createEmpty();
+	const multilineFormat = { type: multilineTag };
 
-	append( tree, '' );
+	let lastCharacterFormats;
+	let lastCharacter;
+
+	// If we're building a multiline tree, start off with a multiline element.
+	if ( multilineTag ) {
+		append( append( tree, { type: multilineTag } ), '' );
+		lastCharacterFormats = [ multilineFormat ];
+	} else {
+		append( tree, '' );
+	}
 
 	for ( let i = 0; i < formatsLength; i++ ) {
 		const character = text.charAt( i );
-		const characterFormats = formats[ i ];
-		const lastCharacterFormats = formats[ i - 1 ];
+		let characterFormats = formats[ i ];
+
+		// Set multiline tags in queue for building the tree.
+		// For list items we need to take into account nested list items.
+		if ( multilineTag === 'li' ) {
+			characterFormats = ( characterFormats || [] ).reduce( ( accumulator, format ) => {
+				if ( format.type === 'ol' || format.type === 'ul' ) {
+					accumulator.push( format );
+					accumulator.push( multilineFormat );
+				} else if ( character !== '\u2028' ) {
```

Comment:
Can we give this magic string a named constant? Out of context, it means nothing to me.